core/src/OpenMP/Kokkos_OpenMP_Instance.hpp

```diff
@@ -163,24 +163,47 @@ inline std::vector<OpenMP> create_OpenMP_instances(
         "Kokkos::abort: Partition not enough resources left to create the last "
         "instance.");
   }
-  instances[weights.size() - 1] = OpenMP(resources_left);
+  pool_sizes[weights.size() - 1] = resources_left;
 
-  return instances;
+  return pool_sizes;
 }
 }  // namespace Impl
 
 template <typename... Args>
-std::vector<OpenMP> partition_space(OpenMP const& main_instance, Args... args) {
+std::array<OpenMP, sizeof...(Args)> partition_space(OpenMP const& main_instance,
+                                                    Args... args) {
   // Unpack the arguments and create the weight vector. Note that if not all of
   // the types are the same, you will get a narrowing warning.
   std::vector<std::common_type_t<Args...>> const weights = {args...};
-  return Impl::create_OpenMP_instances(main_instance, weights);
+
+  // Calculate pool size for each OpenMP instance
+  const auto pool_sizes =
+      Impl::calculate_omp_pool_sizes(main_instance, weights);
+
+  // Create array of OpenMP instances from pool sizes
+  auto create_partitions = [&pool_sizes]<std::size_t... Indices>(
+                               std::index_sequence<Indices...>) {
+    return std::array<OpenMP, sizeof...(Args)>{OpenMP(pool_sizes[Indices])...};
+  };
+
+  return create_partitions(std::make_index_sequence<sizeof...(Args)>{});
```

Comment:
@dalg24 This is the only backend where it wasn't obvious to me that I could use folded expressions. What I've done might not be the most efficiently written, let me know if you have improvements/dislike the approach.